src/subprocess-posix.cc

```diff
@@ -169,11 +170,61 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+bool SubprocessSet::child_exited_;
+
+void SubprocessSet::SetChildExited(int signum) {
+  child_exited_ = true;
+}
+
+void SubprocessSet::HandleChildExit() {
+  if (!child_exited_)
+    return;
+  child_exited_ = false;
+  for (int i = running_.size(); i > 0; --i)
+    WaitChildExit(false);
+}
+
+void SubprocessSet::WaitChildExit(bool block) {
+  int status = 0;
+  int options = block ? 0 : WNOHANG;
+  pid_t pid = waitpid(-1, &status, options);
+  if (pid == -1)
+    Fatal("waitpid: %s", strerror(errno));
```

Comment:
In 'blocking' mode, pressing Ctrl-C will interrupt waitpid()
You should probably return gracefully here if errno set to EINTR to let ninja exit cleanly